main.py

```diff
@@ -172,8 +216,30 @@ def _handle_multipart_flow_stats(self, reply, switch):
                 all_flows = self._multipart_replies_flows.setdefault(switch.id,
                                                                      [])
                 all_flows.extend(flows)
+
+            xid = int(reply.header.xid)
+            if xid in self._multipart_replies_xids.get(switch.id, {}):
+                with self._multipart_replies_xids_lock[switch.id]:
+                    self._multipart_replies_xids[switch.id][xid] -= 1
+
             if reply.flags.value % 2 == 0:  # Last bit means more replies
-                self._update_switch_flows(switch)
+                # make sure no more parts are missing, wait at most half of
```

Comment:
Let me double check if I followed correctly the intention here, @italovalcy :

At this point, once `reply.flags.value == 0`, I believe we have a strong guarantee based on the spec that no more messages are expected, right?  If the counter `_multipart_replies_xids[xid]` doesn't equal to 0 looks like it would mean it's an inconsistent state and data was lost (so Kytos's controller/server mistake), should we update the flows anyway in that case? I'm trying to understand if we need the busy wait and how it would handle this case. I understood that `settings.STATS_INTERVAL/2` would be upper bounded by the `max_delay`  but don't we also need to make sure that the number of multipart replies is decremented to `0` to update? Or otherwise just abort the update (do any potential clean up) and leave this for another execution to try to update in a next thread iteration?

Reply:
Thanks for the question and the opportunity to clarify the adopted approach, @viniarck.

The `reply.flags.value % 2 == 0 ` means that the handler function received a part which is supposed to be the last one. At this point, the switch has sent all the parts and each one of them will be handled by a different thread on Kytos side. Since it is a TCP connection, there will be no packet loss in the network. It is very difficult that we lose any part on the Kytos side as well. However, what can happens is the multiple parts being processed in a different order due to the way threads are scheduled (which, btw, was improved significantly with the [introduction of thread pool on the PR you recently submitted](https://github.com/kytos-ng/kytos/pull/149)). Thus, if a switch A has N flows and it replies the flow_stats with 10 parts (let's say part p1, p2, ..., p10), part p10 will have `reply.flags.value % 2 == 0 ` and it will enter in a busy wait to make sure all the other parts p1, p2, .., p9 has been received. When a part is received, it decrements the xid counter here: https://github.com/italovalcy/of_core/blob/issue_25_issue_27/main.py#L223

The cleanup process you mentioned is done once a new flow_stats request is created, here: https://github.com/italovalcy/of_core/blob/issue_25_issue_27/main.py#L113-L115

The upper bound limit is needed just to guarantee that we will not be waiting forever. In case we reach that upper bound, I believe it is worthy to report on the flow_stats because from my field experience the last messages of a multiple part reply are just padding and does not contains useful information.  But that is something we can definitely discuss better and maybe discard that particular transaction.

The busy wait is indeed necessary, because we should check for missing parts only from the last one. It is running in a separated thread though, so there will be no blocking to other functions. I'm opened to other ideas and improvements.

Reply:
I see. Appreciated the extra info @italovalcy . It looks good to me, the last part of my question was to confirm that indeed messages wouldn't be lost for other socket exception that can happen like `SocketError` or `OSError` (rare though), and then only calling the `self._update_switch_flows(switch)`  if ` self._multipart_replies_xids[switch.id][xid] == 0?` since it's expected that at that time all of them were handled the counter should be at 0, otherwise skip the update?

Reply:
Understood, @viniarck! In regards to the exceptions when calling `self._update_switch_flows(switch)`, the idea of only handling a `KeyError` exception if because we only handle some Switch attributes (e.g., the list of FlowStats received from multipart replies, transaction id, counter of parts, etc). There is no socket interaction at that point. The exception handler was added to cover rare cases where overlapping requests were observed but the missing part was received right after [cleaning up the data struct for a new request](https://github.com/italovalcy/of_core/blob/issue_25_issue_27/main.py#L82). In this case, `_update_switch_flows()` would [fail to obtain the list of received FlowStats](https://github.com/italovalcy/of_core/blob/issue_25_issue_27/main.py#L261).